pilot/pkg/serviceregistry/kube/controller/serviceexportcache.go

```diff
@@ -0,0 +1,139 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/go-multierror"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
+
+	"istio.io/istio/pilot/pkg/features"
+	"istio.io/istio/pilot/pkg/model"
+	kubesr "istio.io/istio/pilot/pkg/serviceregistry/kube"
+	"istio.io/istio/pkg/config/host"
+)
+
+// serviceExportCache provides export state for all services in the cluster.
+type serviceExportCache interface {
+	isExported(namespace, serviceName string) bool
+	HasSynced() bool
+	Sync() error
+}
+
+// newServiceExportCache creates a new serviceExportCache that observes the given cluster.
+func newServiceExportCache(c *Controller) serviceExportCache {
+	if c.opts.EnableMCSServiceDiscovery {
+		sec := &serviceExportCacheImpl{
+			Controller: c,
+			informer:   c.client.MCSApisInformer().Multicluster().V1alpha1().ServiceExports().Informer(),
+		}
+
+		// Create the informer and register callbacks.
+		c.registerHandlers(sec.informer, "ServiceExports", sec.onEvent, nil)
+		return sec
+	}
+
+	// MCS Service discovery is disabled. Use a placeholder cache.
+	return disabledServiceExportCache{}
+}
+
+// serviceExportCache reads ServiceExport resources for a single cluster.
+type serviceExportCacheImpl struct {
+	*Controller
+	informer cache.SharedIndexInformer
+}
+
+func (ec *serviceExportCacheImpl) onEvent(obj interface{}, event model.Event) error {
+	se, ok := obj.(*v1alpha1.ServiceExport)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			return fmt.Errorf("couldn't get object from tombstone %#v", obj)
+		}
+		se, ok = tombstone.Obj.(*v1alpha1.ServiceExport)
+		if !ok {
+			return fmt.Errorf("tombstone contained object that is not a ServiceExport %#v", obj)
+		}
+	}
+
+	switch event {
+	case model.EventAdd, model.EventDelete:
+		// Lookup the service associated with this ServiceExport.
+		svcHost := ec.getHostname(se)
+		svc, _ := ec.GetService(svcHost)
+
+		// Rebuild the endpoints based on the new export state.
+		if svc != nil {
+			endpoints := ec.endpoints.buildIstioEndpointsWithService(svc.Attributes.Name, svc.Attributes.Namespace, svcHost)
+			if features.EnableK8SServiceSelectWorkloadEntries {
+				fep := ec.collectWorkloadInstanceEndpoints(svc)
+				endpoints = append(endpoints, fep...)
+			}
+
+			if len(endpoints) > 0 {
+				ec.opts.XDSUpdater.EDSCacheUpdate(ec.Cluster(), string(svc.Hostname), svc.Attributes.Namespace, endpoints)
+			}
+		}
+
+	default:
+		// Don't care about updates.
+	}
+	return nil
+}
+
+func (ec *serviceExportCacheImpl) isExported(namespace, serviceName string) bool {
+	_, err := ec.client.MCSApis().MulticlusterV1alpha1().ServiceExports(namespace).Get(context.TODO(),
```

Comment:
This *MUST* be changed to use the informer cache, not direct API server calls, or you will DOS the api server and blow up Istiod. Speaking from previous experience..